firebase-firestore/src/main/java/com/google/firebase/firestore/local/QueryEngine.java

```diff
@@ -14,28 +14,224 @@
+    // Sort the documents and re-apply the query filter since previously matching documents do not
+    // necessarily still match the query.
+    ImmutableSortedSet<Document> queryResults =
+        new ImmutableSortedSet<>(Collections.emptyList(), query.comparator());
+    for (Map.Entry<DocumentKey, Document> entry : documents) {
+      Document document = entry.getValue();
+      if (query.matches(document)) {
+        queryResults = queryResults.insert(document);
+      }
+    }
+    return queryResults;
+  }
+
+  /**
+   * Determines if a limit query needs to be refilled from cache, making it ineligible for
+   * index-free execution.
+   *
+   * @param limitType The type of limit query for refill calculation.
+   * @param sortedPreviousResults The documents that matched the query when it was last
+   *     synchronized, sorted by the query's comparator.
+   * @param remoteKeys The document keys that matched the query at the last snapshot.
+   * @param limboFreeSnapshotVersion The version of the snapshot when the query was last
+   *     synchronized.
+   */
+  private boolean needsRefill(
+      Query.LimitType limitType,
+      ImmutableSortedSet<Document> sortedPreviousResults,
+      ImmutableSortedSet<DocumentKey> remoteKeys,
+      SnapshotVersion limboFreeSnapshotVersion) {
+    // The query needs to be refilled if a previously matching document no longer matches.
+    if (remoteKeys.size() != sortedPreviousResults.size()) {
+      return true;
+    }
+
+    // Limit queries are not eligible for index-free query execution if there is a potential that an
+    // older document from cache now sorts before a document that was previously part of the limit.
+    // This, however, can only happen if the document at the edge of the limit goes out of limit. If
+    // a document that is not the limit boundary sorts differently, the boundary of the limit itself
+    // did not change and documents from cache will continue to be "rejected" by this boundary.
+    // Therefore, we can ignore any modifications that don't affect the last document.
+    Document documentAtLimitEdge =
+        limitType == Query.LimitType.LIMIT_TO_FIRST
+            ? sortedPreviousResults.getMaxEntry()
+            : sortedPreviousResults.getMinEntry();
+    if (documentAtLimitEdge == null) {
+      // We don't need to refill the query if there were already no documents.
+      return false;
+    }
+    return documentAtLimitEdge.hasPendingWrites()
+        || documentAtLimitEdge.getVersion().compareTo(limboFreeSnapshotVersion) > 0;
+  }
+
+  private ImmutableSortedMap<DocumentKey, Document> executeFullCollectionScan(Query query) {
+    if (Logger.isDebugEnabled()) {
+      Logger.debug(LOG_TAG, "Using full collection scan to execute query: %s", query.toString());
+    }
+    return localDocumentsView.getDocumentsMatchingQuery(query, IndexOffset.NONE);
+  }
+
+  private ImmutableSortedMap<DocumentKey, Document> mergeResults(
```

Comment:
nit: Either add comment or change method name to include the fact that we're merging results with LDV's new documents since the offset